// algosdk/dist/cjs/makeTxn.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.makeApplicationNoOpTxnFromObject = exports.makeApplicationClearStateTxnFromObject = exports.makeApplicationCloseOutTxnFromObject = exports.makeApplicationOptInTxnFromObject = exports.makeApplicationDeleteTxnFromObject = exports.makeApplicationUpdateTxnFromObject = exports.makeApplicationCreateTxnFromObject = exports.makeApplicationCallTxnFromObject = exports.makeAssetTransferTxnWithSuggestedParamsFromObject = exports.makeAssetFreezeTxnWithSuggestedParamsFromObject = exports.makeAssetDestroyTxnWithSuggestedParamsFromObject = exports.makeAssetConfigTxnWithSuggestedParamsFromObject = exports.makeAssetCreateTxnWithSuggestedParamsFromObject = exports.makeBaseAssetConfigTxn = exports.makeKeyRegistrationTxnWithSuggestedParamsFromObject = exports.makePaymentTxnWithSuggestedParamsFromObject = void 0;
const transaction_js_1 = require("./transaction.js");
const base_js_1 = require("./types/transactions/base.js");
const appAccess_js_1 = require("./appAccess.js");
/**
 * Create a new payment transaction
 *
 * @param options - Payment transaction parameters
 */
function makePaymentTxnWithSuggestedParamsFromObject({ sender, receiver, amount, closeRemainderTo, suggestedParams, note, lease, rekeyTo, }) {
    return new transaction_js_1.Transaction({
        type: base_js_1.TransactionType.pay,
        sender,
        note,
        lease,
        rekeyTo,
        suggestedParams,
        paymentParams: {
            receiver,
            amount,
            closeRemainderTo,
        },
    });
}
exports.makePaymentTxnWithSuggestedParamsFromObject = makePaymentTxnWithSuggestedParamsFromObject;
/**
 * Create a new key registration transaction
 *
 * @param options - Key registration transaction parameters
 */
function makeKeyRegistrationTxnWithSuggestedParamsFromObject({ sender, voteKey, selectionKey, stateProofKey, voteFirst, voteLast, voteKeyDilution, nonParticipation, suggestedParams, note, lease, rekeyTo, }) {
    return new transaction_js_1.Transaction({
        type: base_js_1.TransactionType.keyreg,
        sender,
        note,
        lease,
        rekeyTo,
        suggestedParams,
        keyregParams: {
            voteKey,
            selectionKey,
            stateProofKey,
            voteFirst,
            voteLast,
            voteKeyDilution,
            nonParticipation,
        },
    });
}
exports.makeKeyRegistrationTxnWithSuggestedParamsFromObject = makeKeyRegistrationTxnWithSuggestedParamsFromObject;
/**
 * Base function for creating any type of asset config transaction.
 *
 * @param options - Asset config transaction parameters
 */
function makeBaseAssetConfigTxn({ sender, assetIndex, total, decimals, defaultFrozen, manager, reserve, freeze, clawback, unitName, assetName, assetURL, assetMetadataHash, note, lease, rekeyTo, suggestedParams, }) {
    return new transaction_js_1.Transaction({
        type: base_js_1.TransactionType.acfg,
        sender,
        note,
        lease,
        rekeyTo,
        suggestedParams,
        assetConfigParams: {
            assetIndex,
            total,
            decimals,
            defaultFrozen,
            manager,
            reserve,
            freeze,
            clawback,
            unitName,
            assetName,
            assetURL,
            assetMetadataHash,
        },
    });
}
exports.makeBaseAssetConfigTxn = makeBaseAssetConfigTxn;
/**
 * Create a new asset creation transaction
 *
 * @param options - Asset creation transaction parameters
 */
function makeAssetCreateTxnWithSuggestedParamsFromObject({ sender, total, decimals, defaultFrozen, manager, reserve, freeze, clawback, unitName, assetName, assetURL, assetMetadataHash, note, lease, rekeyTo, suggestedParams, }) {
    return makeBaseAssetConfigTxn({
        sender,
        total,
        decimals,
        defaultFrozen,
        manager,
        reserve,
        freeze,
        clawback,
        unitName,
        assetName,
        assetURL,
        assetMetadataHash,
        note,
        lease,
        rekeyTo,
        suggestedParams,
    });
}
exports.makeAssetCreateTxnWithSuggestedParamsFromObject = makeAssetCreateTxnWithSuggestedParamsFromObject;
/**
 * Create a new asset config transaction. This transaction can be issued by the asset manager to
 * change the manager, reserve, freeze, or clawback address.
 *
 * You must respecify existing addresses to keep them the same; leaving a field blank is the same as
 * turning that feature off for this asset.
 *
 * @param options - Asset modification transaction parameters
 */
function makeAssetConfigTxnWithSuggestedParamsFromObject({ sender, assetIndex, manager, reserve, freeze, clawback, strictEmptyAddressChecking, note, lease, rekeyTo, suggestedParams, }) {
    if (!assetIndex) {
        throw Error('assetIndex must be provided');
    }
    const strictChecking = strictEmptyAddressChecking ?? true;
    if (strictChecking &&
        (manager == null || reserve == null || freeze == null || clawback == null)) {
        throw Error('strictEmptyAddressChecking is enabled, but an address is empty. If this is intentional, set strictEmptyAddressChecking to false.');
    }
    return makeBaseAssetConfigTxn({
        sender,
        assetIndex,
        manager,
        reserve,
        freeze,
        clawback,
        note,
        lease,
        rekeyTo,
        suggestedParams,
    });
}
exports.makeAssetConfigTxnWithSuggestedParamsFromObject = makeAssetConfigTxnWithSuggestedParamsFromObject;
/**
 * Create a new asset destroy transaction. This will allow the asset's manager to remove this asset
 * from the ledger, provided all outstanding assets are held by the creator.
 *
 * @param options - Asset destroy transaction parameters
 */
function makeAssetDestroyTxnWithSuggestedParamsFromObject({ sender, assetIndex, note, lease, rekeyTo, suggestedParams, }) {
    if (!assetIndex) {
        throw Error('assetIndex must be provided');
    }
    return makeBaseAssetConfigTxn({
        sender,
        assetIndex,
        note,
        lease,
        rekeyTo,
        suggestedParams,
    });
}
exports.makeAssetDestroyTxnWithSuggestedParamsFromObject = makeAssetDestroyTxnWithSuggestedParamsFromObject;
/**
 * Create a new asset freeze transaction. This transaction allows the asset's freeze manager to
 * freeze or un-freeze an account, blocking or allowing asset transfers to and from the targeted
 * account.
 *
 * @param options - Asset freeze transaction parameters
 */
function makeAssetFreezeTxnWithSuggestedParamsFromObject({ sender, assetIndex, freezeTarget, frozen, suggestedParams, note, lease, rekeyTo, }) {
    return new transaction_js_1.Transaction({
        type: base_js_1.TransactionType.afrz,
        sender,
        note,
        lease,
        rekeyTo,
        suggestedParams,
        assetFreezeParams: {
            assetIndex,
            freezeTarget,
            frozen,
        },
    });
}
exports.makeAssetFreezeTxnWithSuggestedParamsFromObject = makeAssetFreezeTxnWithSuggestedParamsFromObject;
/**
 * Create a new asset transfer transaction.
 *
 * Special case: to opt into an assets, set amount=0 and sender=receiver.
 *
 * @param options - Asset transfer transaction parameters
 */
function makeAssetTransferTxnWithSuggestedParamsFromObject({ sender, receiver, amount, closeRemainderTo, assetSender, note, assetIndex, suggestedParams, rekeyTo, lease, }) {
    if (!assetIndex) {
        throw Error('assetIndex must be provided');
    }
    return new transaction_js_1.Transaction({
        type: base_js_1.TransactionType.axfer,
        sender,
        note,
        lease,
        rekeyTo,
        suggestedParams,
        assetTransferParams: {
            assetIndex,
            receiver,
            amount,
            assetSender,
            closeRemainderTo,
        },
    });
}
exports.makeAssetTransferTxnWithSuggestedParamsFromObject = makeAssetTransferTxnWithSuggestedParamsFromObject;
/**
 * Base function for creating any application call transaction.
 *
 * @param options - Application call transaction parameters
 */
function makeApplicationCallTxnFromObject({ sender, appIndex, onComplete, appArgs, accounts, foreignApps, foreignAssets, boxes, convertToAccess, holdings, locals, access, approvalProgram, clearProgram, numLocalInts, numLocalByteSlices, numGlobalInts, numGlobalByteSlices, extraPages, rejectVersion, note, lease, rekeyTo, suggestedParams, }) {
    if (onComplete == null) {
        throw Error('onComplete must be provided');
    }
    if (access &&
        (accounts || foreignApps || foreignAssets || boxes || holdings || locals)) {
        throw Error('cannot specify both access and other access fields');
    }
    let access2 = access;
    if (convertToAccess) {
        access2 = (0, appAccess_js_1.foreignArraysToResourceReferences)({
            appIndex,
            accounts,
            foreignApps,
            foreignAssets,
            holdings,
            locals,
            boxes,
        });
    }
    return new transaction_js_1.Transaction({
        type: base_js_1.TransactionType.appl,
        sender,
        note,
        lease,
        rekeyTo,
        suggestedParams,
        appCallParams: {
            appIndex,
            onComplete,
            appArgs,
            // Only pass legacy foreign arrays if access is not provided
            accounts: access2 ? undefined : accounts,
            foreignAssets: access2 ? undefined : foreignAssets,
            foreignApps: access2 ? undefined : foreignApps,
            boxes: access2 ? undefined : boxes,
            access: access2,
            approvalProgram,
            clearProgram,
            numLocalInts,
            numLocalByteSlices,
            numGlobalInts,
            numGlobalByteSlices,
            extraPages,
            rejectVersion,
        },
    });
}
exports.makeApplicationCallTxnFromObject = makeApplicationCallTxnFromObject;
/**
 * Make a transaction that will create an application.
 *
 * @param options - Application creation transaction parameters
 */
function makeApplicationCreateTxnFromObject({ sender, onComplete, appArgs, accounts, foreignApps, foreignAssets, boxes, convertToAccess, holdings, locals, access, approvalProgram, clearProgram, numLocalInts, numLocalByteSlices, numGlobalInts, numGlobalByteSlices, extraPages, note, lease, rekeyTo, suggestedParams, }) {
    if (!approvalProgram || !clearProgram) {
        throw Error('approvalProgram and clearProgram must be provided');
    }
    if (onComplete == null) {
        throw Error('onComplete must be provided');
    }
    return makeApplicationCallTxnFromObject({
        sender,
        appIndex: 0,
        onComplete,
        appArgs,
        accounts,
        foreignApps,
        foreignAssets,
        boxes,
        convertToAccess,
        holdings,
        locals,
        access,
        approvalProgram,
        clearProgram,
        numLocalInts,
        numLocalByteSlices,
        numGlobalInts,
        numGlobalByteSlices,
        extraPages,
        note,
        lease,
        rekeyTo,
        suggestedParams,
    });
}
exports.makeApplicationCreateTxnFromObject = makeApplicationCreateTxnFromObject;
/**
 * Make a transaction that changes an application's approval and clear programs
 *
 * @param options - Application update transaction parameters
 */
function makeApplicationUpdateTxnFromObject({ sender, appIndex, appArgs, accounts, foreignApps, foreignAssets, boxes, convertToAccess, holdings, locals, access, approvalProgram, clearProgram, note, lease, rekeyTo, suggestedParams, }) {
    if (!appIndex) {
        throw Error('appIndex must be provided');
    }
    if (!approvalProgram || !clearProgram) {
        throw Error('approvalProgram and clearProgram must be provided');
    }
    return makeApplicationCallTxnFromObject({
        sender,
        appIndex,
        onComplete: base_js_1.OnApplicationComplete.UpdateApplicationOC,
        appArgs,
        accounts,
        foreignApps,
        foreignAssets,
        boxes,
        convertToAccess,
        holdings,
        locals,
        access,
        approvalProgram,
        clearProgram,
        note,
        lease,
        rekeyTo,
        suggestedParams,
    });
}
exports.makeApplicationUpdateTxnFromObject = makeApplicationUpdateTxnFromObject;
/**
 * Make a transaction that deletes an application
 *
 * @param options - Application deletion transaction parameters
 */
function makeApplicationDeleteTxnFromObject({ sender, appIndex, appArgs, accounts, foreignApps, foreignAssets, boxes, convertToAccess, holdings, locals, access, note, lease, rekeyTo, suggestedParams, }) {
    if (!appIndex) {
        throw Error('appIndex must be provided');
    }
    return makeApplicationCallTxnFromObject({
        sender,
        appIndex,
        onComplete: base_js_1.OnApplicationComplete.DeleteApplicationOC,
        appArgs,
        accounts,
        foreignApps,
        foreignAssets,
        boxes,
        convertToAccess,
        holdings,
        locals,
        access,
        note,
        lease,
        rekeyTo,
        suggestedParams,
    });
}
exports.makeApplicationDeleteTxnFromObject = makeApplicationDeleteTxnFromObject;
/**
 * Make a transaction that opts in to use an application
 *
 * @param options - Application opt-in transaction parameters
 */
function makeApplicationOptInTxnFromObject({ sender, appIndex, appArgs, accounts, foreignApps, foreignAssets, boxes, convertToAccess, holdings, locals, access, note, lease, rekeyTo, suggestedParams, }) {
    if (!appIndex) {
        throw Error('appIndex must be provided');
    }
    return makeApplicationCallTxnFromObject({
        sender,
        appIndex,
        onComplete: base_js_1.OnApplicationComplete.OptInOC,
        appArgs,
        accounts,
        foreignApps,
        foreignAssets,
        boxes,
        note,
        convertToAccess,
        holdings,
        locals,
        access,
        lease,
        rekeyTo,
        suggestedParams,
    });
}
exports.makeApplicationOptInTxnFromObject = makeApplicationOptInTxnFromObject;
/**
 * Make a transaction that closes out a user's state in an application
 *
 * @param options - Application close-out transaction parameters
 */
function makeApplicationCloseOutTxnFromObject({ sender, appIndex, appArgs, accounts, foreignApps, foreignAssets, boxes, convertToAccess, holdings, locals, access, note, lease, rekeyTo, suggestedParams, }) {
    if (!appIndex) {
        throw Error('appIndex must be provided');
    }
    return makeApplicationCallTxnFromObject({
        sender,
        appIndex,
        onComplete: base_js_1.OnApplicationComplete.CloseOutOC,
        appArgs,
        accounts,
        foreignApps,
        foreignAssets,
        boxes,
        convertToAccess,
        holdings,
        locals,
        access,
        note,
        lease,
        rekeyTo,
        suggestedParams,
    });
}
exports.makeApplicationCloseOutTxnFromObject = makeApplicationCloseOutTxnFromObject;
/**
 * Make a transaction that clears a user's state in an application
 *
 * @param options - Application clear state transaction parameters
 */
function makeApplicationClearStateTxnFromObject({ sender, appIndex, appArgs, accounts, foreignApps, foreignAssets, boxes, convertToAccess, holdings, locals, access, note, lease, rekeyTo, suggestedParams, }) {
    if (!appIndex) {
        throw Error('appIndex must be provided');
    }
    return makeApplicationCallTxnFromObject({
        sender,
        appIndex,
        onComplete: base_js_1.OnApplicationComplete.ClearStateOC,
        appArgs,
        accounts,
        foreignApps,
        foreignAssets,
        boxes,
        convertToAccess,
        holdings,
        locals,
        access,
        note,
        lease,
        rekeyTo,
        suggestedParams,
    });
}
exports.makeApplicationClearStateTxnFromObject = makeApplicationClearStateTxnFromObject;
/**
 * Make a transaction that just calls an application, doing nothing on completion
 *
 * @param options - Application no-op transaction parameters
 */
function makeApplicationNoOpTxnFromObject({ sender, appIndex, appArgs, accounts, foreignApps, foreignAssets, boxes, convertToAccess, holdings, locals, access, note, lease, rekeyTo, suggestedParams, }) {
    if (!appIndex) {
        throw Error('appIndex must be provided');
    }
    return makeApplicationCallTxnFromObject({
        sender,
        appIndex,
        onComplete: base_js_1.OnApplicationComplete.NoOpOC,
        appArgs,
        accounts,
        foreignApps,
        foreignAssets,
        boxes,
        convertToAccess,
        holdings,
        locals,
        access,
        note,
        lease,
        rekeyTo,
        suggestedParams,
    });
}
exports.makeApplicationNoOpTxnFromObject = makeApplicationNoOpTxnFromObject;
