// algosdk/dist/cjs/encoding/schema/array.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ArraySchema = void 0;
const encoding_js_1 = require("../encoding.js");
/* eslint-disable class-methods-use-this */
class ArraySchema extends encoding_js_1.Schema {
    constructor(itemSchema) {
        super();
        this.itemSchema = itemSchema;
    }
    defaultValue() {
        return [];
    }
    isDefaultValue(data) {
        return Array.isArray(data) && data.length === 0;
    }
    prepareMsgpack(data) {
        if (Array.isArray(data)) {
            return data.map((item) => this.itemSchema.prepareMsgpack(item));
        }
        throw new Error('ArraySchema data must be an array');
    }
    fromPreparedMsgpack(encoded, rawStringProvider) {
        if (Array.isArray(encoded)) {
            return encoded.map((item, index) => this.itemSchema.fromPreparedMsgpack(item, rawStringProvider.withArrayElement(index)));
        }
        throw new Error(`ArraySchema encoded data must be an array: ${encoded} (${typeof encoded})`);
    }
    prepareJSON(data, options) {
        if (Array.isArray(data)) {
            return data.map((item) => this.itemSchema.prepareJSON(item, options));
        }
        throw new Error('ArraySchema data must be an array');
    }
    fromPreparedJSON(encoded) {
        if (Array.isArray(encoded)) {
            return encoded.map((item) => this.itemSchema.fromPreparedJSON(item));
        }
        throw new Error(`ArraySchema encoded data must be an array: ${encoded} (${typeof encoded})`);
    }
}
exports.ArraySchema = ArraySchema;
