// algosdk/dist/cjs/client/v2/algod/disassemble.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.setHeaders = void 0;
const binarydata_js_1 = require("../../../encoding/binarydata.js");
const encoding_js_1 = require("../../../encoding/encoding.js");
const types_js_1 = require("./models/types.js");
const jsonrequest_js_1 = __importDefault(require("../jsonrequest.js"));
/**
 * Sets the default header (if not previously set)
 * @param headers - A headers object
 */
function setHeaders(headers = {}) {
    let hdrs = headers;
    if (Object.keys(hdrs).every((key) => key.toLowerCase() !== 'content-type')) {
        hdrs = { ...headers };
        hdrs['Content-Type'] = 'text/plain';
    }
    return hdrs;
}
exports.setHeaders = setHeaders;
/**
 * Executes disassemble
 */
class Disassemble extends jsonrequest_js_1.default {
    constructor(c, source) {
        super(c);
        this.source = source;
    }
    // eslint-disable-next-line class-methods-use-this
    path() {
        return `/v2/teal/disassemble`;
    }
    executeRequest(headers, customOptions) {
        const txHeaders = setHeaders(headers);
        return this.c.post({
            relativePath: this.path(),
            data: (0, binarydata_js_1.coerceToBytes)(this.source),
            query: this.query,
            requestHeaders: txHeaders,
            customOptions,
        });
    }
    // eslint-disable-next-line class-methods-use-this
    prepare(response) {
        return (0, encoding_js_1.decodeJSON)(response.getJSONText(), types_js_1.DisassembleResponse);
    }
}
exports.default = Disassemble;
