// algosdk/dist/cjs/client/v2/algod/dryrun.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const encoding_js_1 = require("../../../encoding/encoding.js");
const jsonrequest_js_1 = __importDefault(require("../jsonrequest.js"));
const compile_js_1 = require("./compile.js");
const types_js_1 = require("./models/types.js");
class Dryrun extends jsonrequest_js_1.default {
    constructor(c, dr) {
        super(c);
        this.blob = (0, encoding_js_1.encodeMsgpack)(dr);
    }
    // eslint-disable-next-line class-methods-use-this
    path() {
        return '/v2/teal/dryrun';
    }
    executeRequest(headers, customOptions) {
        const txHeaders = (0, compile_js_1.setHeaders)(headers);
        return this.c.post({
            relativePath: this.path(),
            data: this.blob,
            requestHeaders: txHeaders,
            customOptions,
        });
    }
    // eslint-disable-next-line class-methods-use-this
    prepare(response) {
        return (0, encoding_js_1.decodeJSON)(response.getJSONText(), types_js_1.DryrunResponse);
    }
}
exports.default = Dryrun;
