// algosdk/dist/cjs/client/v2/algod/suggestedParams.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const jsonrequest_js_1 = __importDefault(require("../jsonrequest.js"));
const encoding_js_1 = require("../../../encoding/encoding.js");
const types_js_1 = require("./models/types.js");
/**
 * Returns the common needed parameters for a new transaction, in a format the transaction builder expects
 */
class SuggestedParamsRequest extends jsonrequest_js_1.default {
    /* eslint-disable class-methods-use-this */
    path() {
        return '/v2/transactions/params';
    }
    prepare(response) {
        const params = (0, encoding_js_1.decodeJSON)(response.getJSONText(), types_js_1.TransactionParametersResponse);
        return {
            flatFee: false,
            fee: params.fee,
            firstValid: params.lastRound,
            lastValid: params.lastRound + BigInt(1000),
            genesisID: params.genesisId,
            genesisHash: params.genesisHash,
            minFee: params.minFee,
            consensusVersion: params.consensusVersion,
        };
    }
}
exports.default = SuggestedParamsRequest;
